File: app/src/main/java/me/devsaki/hentoid/views/ZoomableRecyclerView.java
package me.devsaki.hentoid.views;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Point;
import android.util.AttributeSet;
import android.view.HapticFeedbackConstants;
import android.view.MotionEvent;
import android.view.ViewConfiguration;
import android.view.ViewPropertyAnimator;
import android.view.animation.DecelerateInterpolator;

import androidx.annotation.Nullable;
import androidx.core.util.Consumer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.annimon.stream.function.DoubleConsumer;

import me.devsaki.hentoid.util.Helper;
import me.devsaki.hentoid.util.Preferences;
import me.devsaki.hentoid.widget.OnZoneTapListener;
import me.devsaki.hentoid.widget.ViewZoomGestureListener;
import me.devsaki.hentoid.widget.ViewZoomGestureListener.Listener;
import timber.log.Timber;

/**
 * Zoomable RecyclerView that supports gestures
 * To be used inside a {@link ZoomableFrame}
 * <p>
 * Credits go to the Tachiyomi team
 */
public class ZoomableRecyclerView extends RecyclerView {

    private static final long ANIMATOR_DURATION_TIME = 200;
    private static final float DEFAULT_SCALE = 1f;
    private static final float MAX_SCALE = 3f;

    private boolean isZooming = false;
    private boolean isLongTapZooming = false;

    private boolean atFirstPosition = false;
    private boolean atLastPosition = false;

    private int halfWidth = 0;
    private int halfHeight = 0;
    private int firstVisibleItemPosition = 0;
    private int lastVisibleItemPosition = 0;
    private float scale = DEFAULT_SCALE;

    private final GestureListener listener = new GestureListener();
    private DoubleConsumer scaleListener = null;
    private final Detector detector = new Detector(getContext(), listener);
    private Consumer<Point> getMaxDimensionsListener = null;

    private OnZoneTapListener tapListener;
    private LongTapListener longTapListener;

    // Hack to access these values outside of a View
    private Point maxBitmapDimensions = null;

    private boolean longTapZoomEnabled = true;


    @Override
    public void onDraw(Canvas c) {
        super.onDraw(c);
        if (null == maxBitmapDimensions && getMaxDimensionsListener != null) {
            maxBitmapDimensions = new Point(c.getMaximumBitmapWidth(), c.getMaximumBitmapHeight());
            getMaxDimensionsListener.accept(maxBitmapDimensions);
        }
    }

    public void setOnGetMaxDimensionsListener(Consumer<Point> getMaxDimensionsListener) {
        this.getMaxDimensionsListener = getMaxDimensionsListener;
    }


    public ZoomableRecyclerView(Context context) {
        super(context);
    }

    public ZoomableRecyclerView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public ZoomableRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }


    public void setTapListener(OnZoneTapListener tapListener) {
        this.tapListener = tapListener;
    }

    public void setLongTapListener(LongTapListener longTapListener) {
        this.longTapListener = longTapListener;
    }

    public void setOnScaleListener(DoubleConsumer scaleListener) {
        this.scaleListener = scaleListener;
    }

    public void setLongTapZoomEnabled(boolean longTapZoomEnabled) {
        this.longTapZoomEnabled = longTapZoomEnabled;
    }

    public interface LongTapListener {
        boolean onListen(MotionEvent ev);
    }


    @Override
    protected void onMeasure(int widthSpec, int heightSpec) {
        halfWidth = MeasureSpec.getSize(widthSpec) / 2;
        halfHeight = MeasureSpec.getSize(heightSpec) / 2;
        super.onMeasure(widthSpec, heightSpec);
    }


    @Override
    public boolean onTouchEvent(MotionEvent e) {
        detector.onTouchEvent(e);
        return super.onTouchEvent(e);
    }

    @Override
    public void onScrolled(int dx, int dy) {
        super.onScrolled(dx, dy);
        LinearLayoutManager layoutManager = (LinearLayoutManager) getLayoutManager();
        if (layoutManager != null) {
            lastVisibleItemPosition = layoutManager.findLastVisibleItemPosition();
            firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();
        }
    }


    @Override
    public void onScrollStateChanged(int state) {
        super.onScrollStateChanged(state);
        LayoutManager layoutManager = getLayoutManager();
        if (layoutManager != null) {
            int visibleItemCount = layoutManager.getChildCount();
            int totalItemCount = layoutManager.getItemCount();
            atLastPosition = visibleItemCount > 0 && lastVisibleItemPosition == totalItemCount - 1;
            atFirstPosition = firstVisibleItemPosition == 0;
        }
    }

    private float getPositionX(float positionX) {
        float maxPositionX = halfWidth * (scale - 1);
        return Helper.coerceIn(positionX, -maxPositionX, maxPositionX);
    }

    private float getPositionY(float positionY) {
        float maxPositionY = halfHeight * (scale - 1);
        return Helper.coerceIn(positionY, -maxPositionY, maxPositionY);
    }

    public float getScale() {
        return scale;
    }

    public void resetScale() {
        zoom(scale, DEFAULT_SCALE, getX(), 0f, getY(), 0f);
    }

    private void zoom(
            float fromScale,
            float toScale,
            float fromX,
            float toX,
            float fromY,
            float toY
    ) {
        isZooming = true;
        AnimatorSet animatorSet = new AnimatorSet();

        ValueAnimator translationXAnimator = ValueAnimator.ofFloat(fromX, toX);
        translationXAnimator.addUpdateListener(animation -> setX((float) animation.getAnimatedValue()));

        ValueAnimator translationYAnimator = ValueAnimator.ofFloat(fromY, toY);
        translationYAnimator.addUpdateListener(animation -> setY((float) animation.getAnimatedValue()));

        ValueAnimator scaleAnimator = ValueAnimator.ofFloat(fromScale, toScale);
        scaleAnimator.addUpdateListener(animation -> setScaleRate((float) animation.getAnimatedValue()));

        animatorSet.playTogether(translationXAnimator, translationYAnimator, scaleAnimator);
        animatorSet.setDuration(Preferences.isReaderZoomTransitions() ? ANIMATOR_DURATION_TIME : 0);
        animatorSet.setInterpolator(new DecelerateInterpolator());
        animatorSet.start();
        animatorSet.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
                // No need to define any behaviour here
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                isZooming = false;
                scale = toScale;
                if (scaleListener != null) scaleListener.accept(scale);
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                // No need to define any behaviour here
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
                // No need to define any behaviour here
            }
        });
    }

    private boolean canMoveHorizontally() {
        if (getLayoutManager() != null)
            return (getLayoutManager().canScrollVertically() || (getLayoutManager().canScrollHorizontally() && (atFirstPosition || atLastPosition)));
        else return false;
    }

    private boolean canMoveVertically() {
        if (getLayoutManager() != null)
            return (getLayoutManager().canScrollHorizontally() || (getLayoutManager().canScrollVertically() && (atFirstPosition || atLastPosition)));
        else return false;
    }

    boolean zoomFling(int velocityX, int velocityY) {
        if (scale <= DEFAULT_SCALE) return false;

        float distanceTimeFactor = 0.4f;
        Float newX = null;
        Float newY = null;

        if (velocityX != 0 && canMoveHorizontally()) {
            float dx = (distanceTimeFactor * velocityX / 2);
            newX = getPositionX(getX() + dx);
        }
        if (velocityY != 0 && canMoveVertically()) {
            float dy = (distanceTimeFactor * velocityY / 2);
            newY = getPositionY(getY() + dy);
        }

        ViewPropertyAnimator animation = animate();
        if (newX != null) animation.x(newX);
        if (newY != null) animation.y(newY);
        animation.setInterpolator(new DecelerateInterpolator())
                .setDuration(400)
                .start();

        return true;
    }

    private void setScaleRate(float rate) {
        setScaleX(rate);
        setScaleY(rate);
    }

    void onScale(float scaleFactor) {
        scale *= scaleFactor;
        scale = Helper.coerceIn(scale, DEFAULT_SCALE, MAX_SCALE);

        Timber.i(">> scale %s -> %s", scaleFactor, scale);

        setScaleRate(scale);

        if (scale != DEFAULT_SCALE) {
            setX(getPositionX(getX()));
            setY(getPositionY(getY()));
        } else {
            setX(0f);
            setY(0f);
        }

        if (scaleListener != null) scaleListener.accept(scale);
    }

    void onScaleBegin() {
        if (detector.isDoubleTapping) {
            detector.isQuickScaling = true;
        }
    }

    void onScaleEnd() {
        if (getScaleX() < DEFAULT_SCALE) {
            zoom(scale, DEFAULT_SCALE, getX(), 0f, getY(), 0f);
        }
    }

    class GestureListener extends Listener {
        @Override
        public boolean onDoubleTap(MotionEvent e) {
            detector.isDoubleTapping = true;
            return false;
        }

        @Override
        public boolean onSingleTapConfirmed(MotionEvent e) {
            LayoutManager llm = getLayoutManager();
            if (llm instanceof LinearLayoutManager) {
                int orientation = ((LinearLayoutManager) llm).getOrientation();
                if (tapListener != null && orientation == LinearLayoutManager.VERTICAL)
                    tapListener.onSingleTapConfirmedAction(e);
            }
            return false;
        }

        @Override
        public void onLongTapConfirmed(MotionEvent ev) {
            if (longTapListener != null && longTapListener.onListen(ev)) {
                performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
            }
            if (longTapZoomEnabled) longTapZoom(ev);
        }

        private void longTapZoom(MotionEvent ev) {
            if (isZooming) return;

            float toScale = 2f;
            float toX = (halfWidth - ev.getX()) * (toScale - 1);
            float toY = (halfHeight - ev.getY()) * (toScale - 1);
            isLongTapZooming = true;
            zoom(DEFAULT_SCALE, toScale, 0f, toX, 0f, toY);
        }

        @Override
        public void onDoubleTapConfirmed(MotionEvent ev) {
            if (!isZooming) {
                if (getScaleX() != DEFAULT_SCALE) {
                    zoom(scale, DEFAULT_SCALE, getX(), 0f, getY(), 0f);
                } else {
                    float toScale = 2f;
                    float toX = (halfWidth - ev.getX()) * (toScale - 1);
                    float toY = (halfHeight - ev.getY()) * (toScale - 1);
                    zoom(DEFAULT_SCALE, toScale, 0f, toX, 0f, toY);
                }
            }
        }
    }

    class Detector extends ViewZoomGestureListener {

        private int scrollPointerId = 0;
        private int downX = 0;
        private int downY = 0;
        private final int touchSlop = ViewConfiguration.get(getContext()).getScaledTouchSlop();
        private boolean isZoomDragging = false;
        boolean isDoubleTapping = false;
        boolean isQuickScaling = false;

        Detector(Context context, Listener listener) {
            super(context, listener);
        }

        @Override
        public boolean onTouchEvent(MotionEvent ev) {
            int action = ev.getActionMasked();
            int actionIndex = ev.getActionIndex();

            switch (action) {
                case MotionEvent.ACTION_DOWN:
                    motionActionDownLocal(ev);
                    break;
                case MotionEvent.ACTION_POINTER_DOWN:
                    motionActionPointerDown(ev, actionIndex);
                    break;
                case MotionEvent.ACTION_MOVE:
                    return motionActionMoveLocal(ev);
                case MotionEvent.ACTION_UP:
                    motionActionUpLocal(ev);
                    break;
                case MotionEvent.ACTION_CANCEL:
                    motionActionCancel();
                    break;
                default:
                    // Nothing to process as default
            }
            return super.onTouchEvent(ev);
        }

        private void motionActionDownLocal(MotionEvent ev) {
            scrollPointerId = ev.getPointerId(0);
            downX = Math.round(ev.getX() + 0.5f);
            downY = Math.round(ev.getY() + 0.5f);
        }

        private void motionActionPointerDown(MotionEvent ev, int actionIndex) {
            scrollPointerId = ev.getPointerId(actionIndex);
            downX = Math.round(ev.getX(actionIndex) + 0.5f);
            downY = Math.round(ev.getY(actionIndex) + 0.5f);
        }

        private boolean motionActionMoveLocal(MotionEvent ev) {
            if (isDoubleTapping && isQuickScaling) {
                return true;
            }

            int index = ev.findPointerIndex(scrollPointerId);
            if (index < 0) {
                return false;
            }

            int x = Math.round(ev.getX(index) + 0.5f);
            int y = Math.round(ev.getY(index) + 0.5f);
            int dx = (canMoveHorizontally()) ? x - downX : 0;
            int dy = (canMoveVertically()) ? y - downY : 0;

            if (!isZoomDragging && scale > DEFAULT_SCALE) {
                boolean startScroll = false;

                if (Math.abs(dx) > touchSlop) {
                    if (dx < 0) {
                        dx += touchSlop;
                    } else {
                        dx -= touchSlop;
                    }
                    startScroll = true;
                }
                if (Math.abs(dy) > touchSlop) {
                    if (dy < 0) {
                        dy += touchSlop;
                    } else {
                        dy -= touchSlop;
                    }
                    startScroll = true;
                }

                if (startScroll) {
                    isZoomDragging = true;
                }
            }

            if (isZoomDragging) {
                zoomScrollBy(dx, dy);
            }
            return super.onTouchEvent(ev);
        }

        private void motionActionUpLocal(MotionEvent ev) {
            if (isDoubleTapping && !isQuickScaling) {
                listener.onDoubleTapConfirmed(ev);
            }
            if (isLongTapZooming) {
                zoom(scale, DEFAULT_SCALE, getX(), 0f, getY(), 0f);
                isLongTapZooming = false;
            }
            isZoomDragging = false;
            isDoubleTapping = false;
            isQuickScaling = false;
        }

        private void motionActionCancel() {
            isZoomDragging = false;
            isDoubleTapping = false;
            isQuickScaling = false;
        }

        private void zoomScrollBy(int dx, int dy) {
            if (dx != 0) {
                setX(getPositionX(getX() + dx));
            }
            if (dy != 0) {
                setY(getPositionY(getY() + dy));
            }
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/views/ZoomableRecyclerView.kt
package me.devsaki.hentoid.views

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Point
import android.util.AttributeSet
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.view.ViewConfiguration
import android.view.animation.DecelerateInterpolator
import androidx.core.util.Consumer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import me.devsaki.hentoid.util.Helper
import me.devsaki.hentoid.util.Preferences
import me.devsaki.hentoid.widget.OnZoneTapListener
import me.devsaki.hentoid.widget.ViewZoomGestureListenerK
import timber.log.Timber
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Zoomable RecyclerView that supports gestures
 * To be used inside a {@link ZoomableFrame}
 * <p>
 * Credits go to the Tachiyomi team
 */
class ZoomableRecyclerView : RecyclerView {

    private var isZooming = false
    private var isLongTapZooming = false

    private var atFirstPosition = false
    private var atLastPosition = false

    private var halfWidth = 0
    private var halfHeight = 0
    private var firstVisibleItemPosition = 0
    private var lastVisibleItemPosition = 0
    var scale = DEFAULT_SCALE
        private set

    private val listener = GestureListener()
    private var scaleListener: Consumer<Double>? = null
    private val detector = Detector(context, listener)
    private var getMaxDimensionsListener: Consumer<Point>? = null

    private var tapListener: OnZoneTapListener? = null
    private var longTapListener: LongTapListener? = null

    // Hack to access these values outside of a View
    private var maxBitmapDimensions: Point? = null

    private var longTapZoomEnabled = true


    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )


    override fun onDraw(c: Canvas) {
        super.onDraw(c)
        if (null == maxBitmapDimensions && getMaxDimensionsListener != null) {
            maxBitmapDimensions = Point(c.maximumBitmapWidth, c.maximumBitmapHeight)
            getMaxDimensionsListener!!.accept(maxBitmapDimensions)
        }
    }

    fun setOnGetMaxDimensionsListener(getMaxDimensionsListener: Consumer<Point>?) {
        this.getMaxDimensionsListener = getMaxDimensionsListener
    }

    fun setTapListener(tapListener: OnZoneTapListener?) {
        this.tapListener = tapListener
    }

    fun setLongTapListener(longTapListener: LongTapListener?) {
        this.longTapListener = longTapListener
    }

    fun setOnScaleListener(scaleListener: Consumer<Double>?) {
        this.scaleListener = scaleListener
    }

    fun setLongTapZoomEnabled(longTapZoomEnabled: Boolean) {
        this.longTapZoomEnabled = longTapZoomEnabled
    }

    interface LongTapListener {
        fun onListen(ev: MotionEvent?): Boolean
    }


    override fun onMeasure(widthSpec: Int, heightSpec: Int) {
        halfWidth = MeasureSpec.getSize(widthSpec) / 2
        halfHeight = MeasureSpec.getSize(heightSpec) / 2
        super.onMeasure(widthSpec, heightSpec)
    }


    override fun onTouchEvent(e: MotionEvent): Boolean {
        detector.onTouchEvent(e)
        return super.onTouchEvent(e)
    }

    override fun onScrolled(dx: Int, dy: Int) {
        super.onScrolled(dx, dy)
        val layoutManager = layoutManager as LinearLayoutManager?
        if (layoutManager != null) {
            lastVisibleItemPosition = layoutManager.findLastVisibleItemPosition()
            firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()
        }
    }


    override fun onScrollStateChanged(state: Int) {
        super.onScrollStateChanged(state)
        val layoutManager = layoutManager
        if (layoutManager != null) {
            val visibleItemCount = layoutManager.childCount
            val totalItemCount = layoutManager.itemCount
            atLastPosition = visibleItemCount > 0 && lastVisibleItemPosition == totalItemCount - 1
            atFirstPosition = firstVisibleItemPosition == 0
        }
    }

    private fun getPositionX(positionX: Float): Float {
        val maxPositionX = halfWidth * (scale - 1)
        return Helper.coerceIn(positionX, -maxPositionX, maxPositionX)
    }

    private fun getPositionY(positionY: Float): Float {
        val maxPositionY = halfHeight * (scale - 1)
        return Helper.coerceIn(positionY, -maxPositionY, maxPositionY)
    }

    fun resetScale() {
        zoom(scale, DEFAULT_SCALE, x, 0f, y, 0f)
    }

    private fun zoom(
        fromScale: Float,
        toScale: Float,
        fromX: Float,
        toX: Float,
        fromY: Float,
        toY: Float
    ) {
        isZooming = true
        val animatorSet = AnimatorSet()
        val translationXAnimator = ValueAnimator.ofFloat(fromX, toX)
        translationXAnimator.addUpdateListener { animation: ValueAnimator ->
            x = animation.animatedValue as Float
        }
        val translationYAnimator = ValueAnimator.ofFloat(fromY, toY)
        translationYAnimator.addUpdateListener { animation: ValueAnimator ->
            y = animation.animatedValue as Float
        }
        val scaleAnimator = ValueAnimator.ofFloat(fromScale, toScale)
        scaleAnimator.addUpdateListener { animation: ValueAnimator ->
            setScaleRate(animation.animatedValue as Float)
        }
        animatorSet.playTogether(translationXAnimator, translationYAnimator, scaleAnimator)
        animatorSet.duration =
            if (Preferences.isReaderZoomTransitions()) ANIMATOR_DURATION_TIME else 0
        animatorSet.interpolator = DecelerateInterpolator()
        animatorSet.start()
        animatorSet.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                // No need to define any behaviour here
            }

            override fun onAnimationEnd(animation: Animator) {
                isZooming = false
                scale = toScale
                if (scaleListener != null) scaleListener!!.accept(scale.toDouble())
            }

            override fun onAnimationCancel(animation: Animator) {
                // No need to define any behaviour here
            }

            override fun onAnimationRepeat(animation: Animator) {
                // No need to define any behaviour here
            }
        })
    }

    private fun canMoveHorizontally(): Boolean {
        return if (layoutManager != null) layoutManager!!.canScrollVertically() || layoutManager!!.canScrollHorizontally() && (atFirstPosition || atLastPosition) else false
    }

    private fun canMoveVertically(): Boolean {
        return if (layoutManager != null) layoutManager!!.canScrollHorizontally() || layoutManager!!.canScrollVertically() && (atFirstPosition || atLastPosition) else false
    }

    fun zoomFling(velocityX: Int, velocityY: Int): Boolean {
        if (scale <= DEFAULT_SCALE) return false
        val distanceTimeFactor = 0.4f
        var newX: Float? = null
        var newY: Float? = null
        if (velocityX != 0 && canMoveHorizontally()) {
            val dx = distanceTimeFactor * velocityX / 2
            newX = getPositionX(x + dx)
        }
        if (velocityY != 0 && canMoveVertically()) {
            val dy = distanceTimeFactor * velocityY / 2
            newY = getPositionY(y + dy)
        }
        val animation = animate()
        if (newX != null) animation.x(newX)
        if (newY != null) animation.y(newY)
        animation.setInterpolator(DecelerateInterpolator())
            .setDuration(400)
            .start()
        return true
    }

    private fun setScaleRate(rate: Float) {
        scaleX = rate
        scaleY = rate
    }

    fun onScale(scaleFactor: Float) {
        scale *= scaleFactor
        scale = Helper.coerceIn(scale, DEFAULT_SCALE, MAX_SCALE)
        Timber.i(">> scale %s -> %s", scaleFactor, scale)
        setScaleRate(scale)
        if (scale != DEFAULT_SCALE) {
            x = getPositionX(x)
            y = getPositionY(y)
        } else {
            x = 0f
            y = 0f
        }
        if (scaleListener != null) scaleListener!!.accept(scale.toDouble())
    }

    fun onScaleBegin() {
        if (detector.isDoubleTapping) {
            detector.isQuickScaling = true
        }
    }

    fun onScaleEnd() {
        if (scaleX < DEFAULT_SCALE) {
            zoom(scale, DEFAULT_SCALE, x, 0f, y, 0f)
        }
    }

    inner class GestureListener : ViewZoomGestureListenerK.Listener() {
        override fun onDoubleTap(e: MotionEvent): Boolean {
            detector.isDoubleTapping = true
            return false
        }

        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            val llm: LayoutManager? = layoutManager
            if (llm is LinearLayoutManager) {
                val orientation = llm.orientation
                if (orientation == LinearLayoutManager.VERTICAL)
                    tapListener?.onSingleTapConfirmedAction(e)
            }
            return false
        }

        override fun onLongTapConfirmed(ev: MotionEvent) {
            longTapListener?.let {
                if (it.onListen(ev)) performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            }
            if (longTapZoomEnabled) longTapZoom(ev)
        }

        private fun longTapZoom(ev: MotionEvent) {
            if (isZooming) return
            val toScale = 2f
            val toX: Float = (halfWidth - ev.x) * (toScale - 1)
            val toY: Float = (halfHeight - ev.y) * (toScale - 1)
            isLongTapZooming = true
            zoom(DEFAULT_SCALE, toScale, 0f, toX, 0f, toY)
        }

        override fun onDoubleTapConfirmed(ev: MotionEvent) {
            if (!isZooming) {
                if (scaleX != DEFAULT_SCALE) {
                    zoom(scale, DEFAULT_SCALE, x, 0f, y, 0f)
                } else {
                    val toScale = 2f
                    val toX: Float = (halfWidth - ev.x) * (toScale - 1)
                    val toY: Float = (halfHeight - ev.y) * (toScale - 1)
                    zoom(DEFAULT_SCALE, toScale, 0f, toX, 0f, toY)
                }
            }
        }
    }

    inner class Detector(context: Context, listener: Listener) :
        ViewZoomGestureListenerK(context, listener) {
        private var scrollPointerId = 0
        private var downX = 0
        private var downY = 0
        private val touchSlop = ViewConfiguration.get(context).scaledTouchSlop
        private var isZoomDragging = false
        var isDoubleTapping = false
        var isQuickScaling = false
        override fun onTouchEvent(ev: MotionEvent): Boolean {
            val action = ev.actionMasked
            val actionIndex = ev.actionIndex
            when (action) {
                MotionEvent.ACTION_DOWN -> motionActionDownLocal(ev)
                MotionEvent.ACTION_POINTER_DOWN -> motionActionPointerDown(ev, actionIndex)
                MotionEvent.ACTION_MOVE -> return motionActionMoveLocal(ev)
                MotionEvent.ACTION_UP -> motionActionUpLocal(ev)
                MotionEvent.ACTION_CANCEL -> motionActionCancel()
                else -> {}
            }
            return super.onTouchEvent(ev)
        }

        private fun motionActionDownLocal(ev: MotionEvent) {
            scrollPointerId = ev.getPointerId(0)
            downX = (ev.x + 0.5f).roundToInt()
            downY = (ev.y + 0.5f).roundToInt()
        }

        private fun motionActionPointerDown(ev: MotionEvent, actionIndex: Int) {
            scrollPointerId = ev.getPointerId(actionIndex)
            downX = (ev.getX(actionIndex) + 0.5f).roundToInt()
            downY = (ev.getY(actionIndex) + 0.5f).roundToInt()
        }

        private fun motionActionMoveLocal(ev: MotionEvent): Boolean {
            if (isDoubleTapping && isQuickScaling) {
                return true
            }
            val index = ev.findPointerIndex(scrollPointerId)
            if (index < 0) {
                return false
            }
            val x = (ev.getX(index) + 0.5f).roundToInt()
            val y = (ev.getY(index) + 0.5f).roundToInt()
            var dx = if (canMoveHorizontally()) x - downX else 0
            var dy = if (canMoveVertically()) y - downY else 0
            if (!isZoomDragging && scale > DEFAULT_SCALE) {
                var startScroll = false
                if (abs(dx) > touchSlop) {
                    if (dx < 0) {
                        dx += touchSlop
                    } else {
                        dx -= touchSlop
                    }
                    startScroll = true
                }
                if (abs(dy) > touchSlop) {
                    if (dy < 0) {
                        dy += touchSlop
                    } else {
                        dy -= touchSlop
                    }
                    startScroll = true
                }
                if (startScroll) {
                    isZoomDragging = true
                }
            }
            if (isZoomDragging) {
                zoomScrollBy(dx, dy)
            }
            return super.onTouchEvent(ev)
        }

        private fun motionActionUpLocal(ev: MotionEvent) {
            if (isDoubleTapping && !isQuickScaling) {
                listener.onDoubleTapConfirmed(ev)
            }
            if (isLongTapZooming) {
                zoom(scale, DEFAULT_SCALE, x, 0f, y, 0f)
                isLongTapZooming = false
            }
            isZoomDragging = false
            isDoubleTapping = false
            isQuickScaling = false
        }

        private fun motionActionCancel() {
            isZoomDragging = false
            isDoubleTapping = false
            isQuickScaling = false
        }

        private fun zoomScrollBy(dx: Int, dy: Int) {
            if (dx != 0) {
                x = getPositionX(x + dx)
            }
            if (dy != 0) {
                y = getPositionY(y + dy)
            }
        }
    }

    companion object {
        const val ANIMATOR_DURATION_TIME: Long = 200
        const val DEFAULT_SCALE = 1f
        const val MAX_SCALE = 3f
    }
}